File: SourceWrappers.Twitter/Twitter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using Tweetinvi;
using Tweetinvi.Models;
using Tweetinvi.Models.Entities;
using Tweetinvi.Parameters;

namespace SourceWrappers.Twitter {
	public class TwitterPostWrapper : IPostWrapper, IStatusUpdate, IDeletable {
		public readonly ITweet _tweet;
		public readonly IMediaEntity _media;
		private ITwitterCredentials _credentials;

		public TwitterPostWrapper(ITweet tweet, IMediaEntity media, ITwitterCredentials credentials) {
			_tweet = tweet;
			_media = media;
			_credentials = credentials;
		}

		public string Title => "";
		public string HTMLDescription {
			get {
				string text = _media == null
					? _tweet.FullText
					: _tweet.FullText.Replace(_media.URL, "");
				return "<p>" + WebUtility.HtmlEncode(text).Replace("\n", "<br/>") + "</p>";
			}
		}
		public bool Mature => _tweet.PossiblySensitive;
		public bool Adult => false;
		public IEnumerable<string> Tags => _tweet.Hashtags.Select(h => h.Text);
		public DateTime Timestamp => _tweet.CreatedAt;
		public string ViewURL => _tweet.Url;
		public string ImageURL => _media != null ? $"{_media.MediaURLHttps}:large" : _tweet.CreatedBy.ProfileImageUrl;
		public string ThumbnailURL => _media != null ? $"{_media.MediaURLHttps}:thumb" : _tweet.CreatedBy.ProfileImageUrl;

		public string SiteName => "Twitter";

		public string FullHTML => HTMLDescription;
		public bool PotentiallySensitive => Mature || Adult;
		public bool HasPhoto => _media != null;
		public IEnumerable<string> AdditionalLinks => _tweet.Entities.Urls.Select(u => u.ExpandedURL);

		public Task DeleteAsync() {
			return Auth.ExecuteOperationWithCredentials(_credentials, async () => {
				await TweetAsync.DestroyTweet(_tweet);
			});
		}
	}

	public class TwitterSourceWrapper : IPagedSourceWrapper<long> {
		private readonly ITwitterCredentials _credentials;
		private readonly bool _photosOnly;

		public TwitterSourceWrapper(ITwitterCredentials credentials, bool photosOnly = true) {
			_credentials = credentials;
			_photosOnly = photosOnly;
		}

		private IUser _user;

		public async Task<IUser> GetUserAsync() {
			if (_user == null) {
				_user = await Auth.ExecuteOperationWithCredentials(_credentials, () => UserAsync.GetAuthenticatedUser());
				if (_user == null) throw new Exception("No user information returned from Twitter (rate limit reached or credentials no longer valid?)");
			}
			return _user;
		}

		public string Name => _photosOnly ? "Twitter (photos)" : "Twitter (text + photos)";
		public int SuggestedBatchSize => 20;

		public Task<FetchResult<long>> MoreAsync(long cursor, int take) {
			return Auth.ExecuteOperationWithCredentials(_credentials, async () => {
				var ps = new UserTimelineParameters {
					ExcludeReplies = false,
					IncludeEntities = true,
					IncludeRTS = true,
					MaxId = cursor,
					MaximumNumberOfTweetsToRetrieve = take
				};
				
				var tweets = await TimelineAsync.GetUserTimeline(await GetUserAsync(), ps);

				IEnumerable<IPostWrapper> Wrap() {
					foreach (var t in tweets.OrderByDescending(t => t.CreatedAt)) {
						if (t.IsRetweet) continue;

						foreach (var m in t.Media) {
							if (m.MediaType == "photo") {
								yield return new TwitterPostWrapper(t, m, _credentials);
							}
						}

						if (!_photosOnly && !t.Media.Any(m => m.MediaType == "photo")) {
							yield return new TwitterPostWrapper(t, null, _credentials);
						}
					}
				}

				return new FetchResult<long>(
					posts: Wrap(),
					next: tweets.Select(t => t.Id).Min() - 1,
					hasMore: tweets.Any());
			});
		}

		public Task<FetchResult<long>> StartAsync(int take) {
			return MoreAsync(-1, take);
		}

		public async Task<IEnumerable<IPostWrapper>> FetchAllAsync(int limit) {
			var list = new List<IPostWrapper>();

			var result = await StartAsync(limit);
			list.AddRange(result.Posts);

			while (result.HasMore && list.Count < limit) {
				result = await MoreAsync(result.Next, limit - list.Count);
				list.AddRange(result.Posts);
			}

			return list;
		}

		public async Task<string> WhoamiAsync() {
			return (await GetUserAsync()).ScreenName;
		}

		public async Task<string> GetUserIconAsync(int size) {
			return (await GetUserAsync()).ProfileImageUrl;
		}
	}
}

File: SourceWrappers/Twitter.fs
namespace SourceWrappers

open ArtworkSourceSpecification
open System.Net
open Tweetinvi.Models
open Tweetinvi.Models.Entities
open Tweetinvi
open Tweetinvi.Parameters
open FSharp.Control
open System

type TwitterPostWrapper(tweet: ITweet, media: IMediaEntity option, twitterClient: ITwitterClient) =
    interface IPostBase with
        member this.Title = ""
        member this.HTMLDescription =
            let text =
                match media with
                | Some m -> tweet.FullText.Replace(m.URL, "")
                | None -> tweet.FullText
            sprintf "<p>%s</p>" (WebUtility.HtmlEncode(text).Replace("\n", "<br/>"))
        member this.Mature = tweet.PossiblySensitive
        member this.Adult = false
        member this.Tags = tweet.Hashtags |> Seq.map (fun t -> t.Text)
        member this.Timestamp = tweet.CreatedAt.UtcDateTime
        member this.ViewURL = tweet.Url
    interface IDeletable with
        member this.SiteName = "Twitter"
        member this.DeleteAsync () = twitterClient.Tweets.DestroyTweetAsync tweet

type TwitterPhotoPostWrapper(tweet: ITweet, media: IMediaEntity, twitterClient: ITwitterClient) =
    inherit TwitterPostWrapper(tweet, Some media, twitterClient)
    interface IRemotePhotoPost with
        member this.ImageURL = sprintf "%s:large" media.MediaURLHttps
        member this.ThumbnailURL = sprintf "%s:thumb" media.MediaURLHttps

type TwitterAnimatedGifPostWrapper(tweet: ITweet, media: IMediaEntity, twitterClient: ITwitterClient) =
    inherit TwitterPostWrapper(tweet, Some media, twitterClient)
    interface IRemoteVideoPost with
        member this.VideoURL =
            media.VideoDetails.Variants
            |> Seq.where (fun v -> v.ContentType = "video/mp4")
            |> Seq.sortByDescending (fun v -> v.Bitrate)
            |> Seq.map (fun v -> v.URL)
            |> Seq.head
        member this.ThumbnailURL = media.MediaURLHttps

type TwitterSourceWrapper(twitterClient: ITwitterClient, photosOnly: bool) =
    inherit AsyncSeqWrapper()

    let user_task = lazy(twitterClient.Users.GetAuthenticatedUserAsync())

    let getUser = user_task.Force() |> Async.AwaitTask

    let wrap t = TwitterPostWrapper (t, None, twitterClient) :> IPostBase
    let wrapPhoto t m = TwitterPhotoPostWrapper (t, m, twitterClient) :> IPostBase
    let wrapAnimatedGif t m = TwitterAnimatedGifPostWrapper (t, m, twitterClient) :> IPostBase

    let isPhoto (m: IMediaEntity) = m.MediaType = "photo"
    let isAnimatedGif (m: IMediaEntity) = m.MediaType = "animated_gif"

    member val BatchSize: int = 20 with get, set

    override __.Name =
        if photosOnly then
            "Twitter (images)"
        else
            "Twitter (text + images)"

    override this.FetchSubmissionsInternal() = asyncSeq {
        let mutable maxId = Nullable<int64>()
        let mutable more = true

        while more do
            let! user = getUser

            let ps = new GetUserTimelineParameters(user)
            ps.ExcludeReplies <- false
            ps.IncludeEntities <- true
            ps.IncludeRetweets <- true
            ps.MaxId <- maxId
            ps.PageSize <- this.BatchSize

            let! tweets = twitterClient.Timelines.GetUserTimelineAsync ps |> Async.AwaitTask

            for t in tweets do
                if not t.IsRetweet then
                    let photos = t.Media |> Seq.filter isPhoto
                    let gifs =
                        if photosOnly
                        then Seq.empty
                        else t.Media |> Seq.filter isAnimatedGif
                    if Seq.isEmpty photos && Seq.isEmpty gifs then
                        if not photosOnly then
                            yield wrap t
                    else
                        for m in photos do
                            yield wrapPhoto t m
                        for m in gifs do
                            yield wrapAnimatedGif t m
                            
            more <- not (Seq.isEmpty tweets)
            if more then
                maxId <- Nullable ((tweets |> Seq.map (fun t -> t.Id) |> Seq.min) - 1L)
    }

    override this.FetchUserInternal() = async {
        let! u = getUser
        return {
            username = u.ScreenName
            icon_url = u.ProfileImageUrl |> Option.ofObj
        }
    }